src/test/java/com/commercetools/sync/sdk2/products/impl/ProductTransformServiceImplTest.java

```diff
@@ -0,0 +1,511 @@
+import static org.mockito.Mockito.*;
+
+import com.commercetools.api.client.ByProjectKeyCustomObjectsGet;
+import com.commercetools.api.client.ByProjectKeyGraphqlPost;
+import com.commercetools.api.client.ProjectApiRoot;
+import com.commercetools.api.defaultconfig.ApiRootBuilder;
+import com.commercetools.api.models.category.CategoryReference;
+import com.commercetools.api.models.custom_object.CustomObject;
+import com.commercetools.api.models.custom_object.CustomObjectPagedQueryResponse;
+import com.commercetools.api.models.graph_ql.GraphQLRequest;
+import com.commercetools.api.models.graph_ql.GraphQLRequestBuilder;
+import com.commercetools.api.models.graph_ql.GraphQLResponse;
+import com.commercetools.api.models.graph_ql.GraphQLResponseBuilder;
+import com.commercetools.api.models.product.ProductDraft;
+import com.commercetools.api.models.product.ProductProjection;
+import com.commercetools.api.models.product.ProductReference;
+import com.commercetools.api.models.product_type.ProductTypeReference;
+import com.commercetools.sync.sdk2.commons.exceptions.ReferenceTransformException;
+import com.commercetools.sync.sdk2.commons.utils.CaffeineReferenceIdToKeyCacheImpl;
+import com.commercetools.sync.sdk2.commons.utils.ReferenceIdToKeyCache;
+import com.commercetools.sync.sdk2.products.service.ProductTransformService;
+import com.commercetools.sync.sdk2.products.service.impl.ProductTransformServiceImpl;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.spotify.futures.CompletableFutures;
+import io.sphere.sdk.client.BadGatewayException;
+import io.vrap.rmf.base.client.ApiHttpMethod;
+import io.vrap.rmf.base.client.ApiHttpResponse;
+import io.vrap.rmf.base.client.utils.json.JsonUtils;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings("unchecked")
+class ProductTransformServiceImplTest {
+
+  final ReferenceIdToKeyCache referenceIdToKeyCache = new CaffeineReferenceIdToKeyCacheImpl();
+
+  @AfterEach
+  void clearCache() {
+    referenceIdToKeyCache.clearCache();
+  }
+
+  @Test
+  void transform_WithAttributeReferences_ShouldReplaceAttributeReferenceIdsWithKeys()
+      throws Exception {
+    // preparation
+    final ProjectApiRoot sourceClient = mock(ProjectApiRoot.class);
+    final ProductTransformService productTransformService =
+        new ProductTransformServiceImpl(sourceClient, referenceIdToKeyCache);
+    final List<ProductProjection> productPage = asList(createProductFromJson("product-key-4.json"));
+
+    String jsonStringProducts =
```

Comment:
You can add `final` in this class.

Reply:
[17b5db3](https://github.com/commercetools/commercetools-sync-java/pull/941/commits/17b5db30f0c9c0799a788f45f54f743d03f9a01b)